training/src/main/scala/com/airbnb/aerosolve/training/photon/ml/data/PhotonMLUtils.scala

```diff
@@ -0,0 +1,198 @@
+
+
+  /**
+    * Create a TrainingExampleAvro schema used by Photon ML:
+    * [[https://github.com/linkedin/photon-ml#input-data-format]]
+    *
+    * @param name              optionally provide a record name for the schema. This is useful when multiple
+    *                          Avro schemas are created and they needed to be distinguished.
+    * @param featureNamespaces if an array of namespaces are provided, we will create a list of
+    *                          extra feature arrays accordingly.
+    * @return an Avro schema instance
+    */
+  def makeTrainingExampleAvroSchema(name: String = "TrainingExampleAvro",
+                                    featureNamespaces: Array[String] = Array[String]()): Schema = {
+    // Define feature item schema
+    val featureItemSchema = SchemaBuilder.record("FeatureItem")
+      .namespace(AVRO_NAMESPACE)
+      .fields()
+      .name(AVRO_NAME).`type`().stringType().noDefault()
+      .name(AVRO_TERM).`type`().stringType().stringDefault("")
+      .name(AVRO_VALUE).`type`().doubleType().noDefault()
+      .endRecord()
+
+    // Define overall training record schema
+    val builder = SchemaBuilder.record(name)
+      .namespace(AVRO_NAMESPACE)
+      .fields()
+      .name(AVRO_UID).`type`().nullable().stringType().noDefault()
+      .name(AVRO_LABEL).`type`().doubleType().noDefault()
+      .name(AVRO_META_DATA_MAP).`type`().map().values().longType().noDefault()
+
+    // Global feature space, important to trigger noDefault() in the end otherwise, the field
+    // will not be successfully committed into the builder
+    builder.name(AVRO_FEATURES).`type`().array().items(featureItemSchema).noDefault()
+
+    // Optionally, if feature family is provided, create per feature family feature space
+    featureNamespaces.foreach { case namespace =>
+      builder.name(makeAvroFeaturesFieldName(namespace))
+        .`type`().array().items(featureItemSchema).noDefault()
+    }
+
+    builder.endRecord()
+  }
+
+  /**
+    * Create an Avro feature record
+    *
+    * @param schema feature item schema, necessary to provide since schemas with different names or
+    *               namespaces are considered to be different.
+    * @param name   feature name
+    * @param value  feature value
+    * @param term   feature term (optionally to be empty)
+    * @return the Avro feature item
+    */
+  def makeAvroFeatureItemRecord(schema: Schema,
+                                name: String,
+                                value: Double,
+                                term: String = ""): GenericRecord = {
+    val r = new GenericData.Record(schema)
+    r.put(AVRO_NAME, name)
+    r.put(AVRO_TERM, term)
+    r.put(AVRO_VALUE, value)
+    r
+  }
+
+  /**
+    * Returns the avro feature array field name for a certain feature family
+    *
+    * @param familyName
+    */
+  def makeAvroFeaturesFieldName(familyName: String): String = familyName.toLowerCase + AVRO_FEATURES.capitalize
```

Comment:
Yes, but it's actually `familynameFeatures`, `capitalize` will move the 1st char to uppercase.